lens/util.go

```diff
@@ -0,0 +1,67 @@
+package lens
+
+import (
+	"context"
+	"errors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/lotus/chain/state"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/node/impl/full"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"golang.org/x/xerrors"
+)
+
+var logger = logging.Logger("visor/lens/lotus")
+
+// OptimizedStateGetActorWithFallback is a helper to obtain an actor in the
+// state of the current tipset without recomputing the full tipset. It does
+// this by obtaining the child tipset (current height+1) and using the
+// pre-computed ParentState().
+//
+// TODO: Remove. See:  https://github.com/filecoin-project/sentinel-visor/issues/196
+func OptimizedStateGetActorWithFallback(ctx context.Context, store *store.ChainStore, fallback full.StateModuleAPI, actor address.Address, tsk types.TipSetKey) (*types.Actor, error) {
+	act, err := efficientStateGetActorFromChainStore(ctx, store, actor, tsk)
+	if err != nil {
+		logger.Warnf("Optimized StateGetActorError: %s. Falling back to default StateGetActor().")
+		return fallback.StateGetActor(ctx, actor, tsk)
+	}
+	return act, nil
+}
+
+func efficientStateGetActorFromChainStore(ctx context.Context, store *store.ChainStore, actor address.Address, tsk types.TipSetKey) (*types.Actor, error) {
```

Comment:
It's unfortunate that this is going to end up going across the API boundary several times (load tipset, load child tipset, loadstatetree (expensive)).
we should think about if there's a way to cache some or any of these things before the API boundary.